segmentstore/server/src/test/java/io/pravega/segmentstore/server/DataRecoveryTestUtils.java

```diff
@@ -0,0 +1,188 @@
+     * @param containerCount    Total number of segment containers.
+     * @return                  A map of lists containing segments by container Ids.
+     * @throws                  IOException in case of exception during the execution.
+     */
+    public static Map<Integer, List<SegmentProperties>> listAllSegments(Storage tier2, int containerCount) throws IOException {
+        SegmentToContainerMapper segToConMapper = new SegmentToContainerMapper(containerCount);
+        Map<Integer, List<SegmentProperties>> segmentToContainers = new HashMap<Integer, List<SegmentProperties>>();
+        log.info("Generating container files with the segments they own...");
+        Iterator<SegmentProperties> it = tier2.listSegments();
+        if (it == null) {
+            return segmentToContainers;
+        }
+        // Iterate through all segments. Put each one of them in its respective list.
+        while (it.hasNext()) {
+            SegmentProperties curr = it.next();
+            int containerId = segToConMapper.getContainerId(curr.getName());
+            List<SegmentProperties> segmentsList = segmentToContainers.get(containerId);
+            if (segmentsList == null) {
+                segmentsList = new ArrayList<>();
+                segmentsList.add(curr);
+                segmentToContainers.put(containerId, segmentsList);
+            } else {
+                segmentToContainers.get(containerId).add(curr);
+            }
+        }
+        return segmentToContainers;
+    }
+
+    public static ScheduledExecutorService createExecutorService(int threadPoolSize) {
+        ScheduledThreadPoolExecutor es = new ScheduledThreadPoolExecutor(threadPoolSize);
+        es.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
+        es.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        es.setRemoveOnCancelPolicy(true);
+        return es;
+    }
+
+     /**
+     * Creates all segments given in the list with the given DebugStreamSegmentContainer.
+     */
+     public static class Worker implements Runnable {
+        private final int containerId;
+        private final DebugStreamSegmentContainer container;
+        private final List<SegmentProperties> segments;
+        public Worker(DebugStreamSegmentContainer container, List<SegmentProperties> segments) {
+            this.container = container;
+            this.containerId = container.getId();
+            this.segments = segments;
+        }
+
+        @Override
+        public void run() {
+            if (segments == null) {
+                return;
+            }
+            log.info("Recovery started for container = {}", containerId);
+            ContainerTableExtension ext = container.getExtension(ContainerTableExtension.class);
+            AsyncIterator<IteratorItem<TableKey>> it = ext.keyIterator(getMetadataSegmentName(containerId),
+                    IteratorArgs.builder().fetchTimeout(TIMEOUT).build()).join();
```

Comment:
`join` will not time out. You have a timeout defined somewhere in this class. Use `get(timeoutMillis, TimeUnit.MILLISECONDS)`

Reply:
And below

Reply:
Done.